### PR TITLE
Fix incorrect backup key format in SSSS

### PR DESCRIPTION
**Pull request by uhoreg**
fixes https://github.com/vector-im/riot-web/issues/12970

I'm planning on tidying up the unit tests a bit, but otherwise this should be ready for review.

### CODE REVIEW

File: src/crypto/index.js [hunk 1]
```diff
@@ -501,172 +509,220 @@ Crypto.prototype.bootstrapSecretStorage = async function({
...
     try {
+        const oldSSSSKey = await this.getSecretStorageKey();
+        const [oldKeyId, oldKeyInfo] = oldSSSSKey || [null, null];
         const decryptionKeys =
               await this._crossSigningInfo.isStoredInSecretStorage(this._secretStorage);
         const inStorage = !setupNewSecretStorage && decryptionKeys;
-        if (decryptionKeys && !(Object.values(decryptionKeys).some(
+
+        if (!decryptionKeys && !keyBackupInfo) {
+            // either we don't have anything, or we've been asked to restart
```
Comment (jryans) [inline src/crypto/index.js:572]: Isn't the "asked to restart" case one where `setupNewSecretStorage === true` and thus `inStorage === false`?

File: src/crypto/index.js [hunk 2]
```diff
@@ -711,8 +767,26 @@ Crypto.prototype.bootstrapSecretStorage = async function({
...
+                    "m.megolm_backup.v1", fixedBackupKey, [newKeyId || oldKeyId],
+                );
+            }
+            const decodedBackupKey = new Uint8Array(olmlib.decodeBase64(
+                fixedBackupKey || sessionBackupKey,
+            ));
+            await this.storeSessionBackupPrivateKey(decodedBackupKey);
+        }
+
+        if (setupNewKeyBackup && !keyBackupInfo) {
```
Comment (jryans) [inline src/crypto/index.js:784]: I recently moved this on develop to fix the key backup cache (https://github.com/matrix-org/matrix-js-sdk/pull/1303/commits/060549656e7f4806496daf7113c9e8144f9e868f), and it looks like you have it in the new position several blocks above, so this block can be removed...?

File: src/crypto/index.js [hunk 3]
```diff
@@ -501,172 +509,220 @@ Crypto.prototype.bootstrapSecretStorage = async function({
...
+
+            // create new cross-signing keys
+            await resetCrossSigning();
+
+            // create a new SSSS key and use the backup key as the new SSSS key
+            const opts = {};
+
+            if (
+                keyBackupInfo.auth_data.private_key_salt &&
+                    keyBackupInfo.auth_data.private_key_iterations
```
Comment (jryans) [inline src/crypto/index.js:616]: Our current trend is to keep multi-line conditionals like this at the same indent on all lines, but probably we should try to automate / lint for this...

```suggestion
                keyBackupInfo.auth_data.private_key_iterations
```

File: src/crypto/index.js [hunk 4]
```diff
@@ -501,172 +509,220 @@ Crypto.prototype.bootstrapSecretStorage = async function({
...
         };
 
+    // create a new SSSS key and set it as default
+    const createSSSS = async (opts, privateKey) => {
+        opts = opts || {};
+        if (privateKey) {
+            opts.key = privateKey;
+        }
+
+        const newKeyId = await this.addSecretStorageKey(
```
Comment (jryans) [inline src/crypto/index.js:519]: This shadows the other `newKeyId` in the outer scope, could we rename one of them?

File: src/crypto/index.js [hunk 5]
```diff
@@ -501,172 +509,220 @@ Crypto.prototype.bootstrapSecretStorage = async function({
...
+            logger.log(
+                "Cross-signing private keys not found in secret storage, " +
+                    "creating new keys",
+            );
+
+            await resetCrossSigning();
+
+            if (oldKeyInfo && oldKeyInfo.algorithm === SECRET_STORAGE_ALGORITHM_V1_AES) {
+                // if we already have a default SSSS key; use it
+                newKeyId = oldKeyId;
```
Comment (jryans) [inline src/crypto/index.js:583]: Is this actually needed? It seems confusing to write to `newKeyId` when it's not actually new, and also all the places that read `newKeyId` seem to either have just generated a new key on the previous line or they check `newKeyId || oldKeyId`, which should cover this case.

File: src/client.js [hunk 6]
```diff
@@ -1815,7 +1815,17 @@ MatrixClient.prototype.restoreKeyBackupWithPassword = async function(
...
 ) {
-    const privKey = decodeBase64(await this.getSecret("m.megolm_backup.v1"));
+    const storedKey = await this.getSecret("m.megolm_backup.v1");
+
+    // ensure that the key is in the right format.  If not, fix the key and
+    // store the fixed version
+    const fixedKey = fixBackupKey(storedKey);
+    if (fixedKey) {
+        const [keyId] = await this.getSecretStorageKey();
+        await this.storeSecret("m.megolm_backup.v1", [keyId]);
```
[the remark this answers is not in the document]
Comment (uhoreg) [reply src/client.js:1825]: Yes.  I managed to fix the places in `crypto/index.js` where I made that mistake, and forgot to fix this one.

File: src/crypto/index.js (hunk 2 above)
Comment (uhoreg) [reply src/crypto/index.js:784]: Hmm, yes.  `git stash` seems to have done something odd here.

File: src/crypto/index.js (hunk 1 above)
Comment (uhoreg) [reply src/crypto/index.js:572]: Yes.  I think at one point, the code matched the comment...

**Comment by uhoreg** [comment]
Manually testing with an actual Riot, I found one bug in the bootstrap process that prevented bootstrapping when starting with only key backup, which is now fixed in this PR.  I also have one scenario in which key backups don't seem to be working at all; I'm investigating whether it's due to this PR or something else.

**Comment by uhoreg** [comment]
> I also have one scenario in which key backups don't seem to be working at all; I'm investigating whether it's due to this PR or something else.

FTR, it appears to be https://github.com/vector-im/riot-web/issues/13151

**Pull request merged**